### ISSUE TITLE
BCash transaction does not sign correctly - broadcast error: signature must be zero for failed check(multi)sig operation

### ISSUE THREAD

**Issue by cucaraca** [MetacoSA/NBitcoin#664]
I have a P2SH transaction that I am trying to spend but broadcasting my resulting signed raw tx to the network gives me an error **16: mandatory-script-verify-flag-failed (signature must be zero for failed check(multi)sig operation)**

**This is how i created the address that received the TX:**

```c#
Network Network => BCash.Instance.Testnet;

public string GetReceiveAddr(uint index)
{
      Mnemonic mnemonic = new Mnemonic(this.Wallet.Mnemonic, Wordlist.English);
      ExtKey masterPriv = mnemonic.DeriveExtKey(this.Wallet.Password);
      ExtPubKey masterPub = masterPriv.Neuter();

      BitcoinAddress addr = masterPub.Derive(this.Wallet.ReceiveChain).Derive(index).ScriptPubKey.Hash.GetAddress(this.Network);

         return addr.ToString();
}
```

**To this address I have received the below tx:**

```json
{
  "hex": "02000000014235fe143004011e387a0e9f88e44c90b6cb84c4d3acb30e73214639aab91625000000006a473044022056fe585e8c79441c491c15ca0be9181cb79eea0bbf6f0971fef50d652130e1790220757f32c237289ea846027b9796e33fd97fbb6df6484969800e977dc531b9b483412103c346eee77a77a8d3e073dacc0532ca7a5b9747bc06d88bf091cac9f4bc8bb792ffffffff02021c9bcf050000001976a91436d2f27bbd826a86db1e93618ce3de89ef33169388ac809698000000000017a9148024f2cc52373aef482b343bc69b83e6bc031a9c8700000000",
  "txid": "93f7c14688324b0ec296f18aef09a4c42bc2143f6d8bdb8179b31318335d0e8e",
  "hash": "93f7c14688324b0ec296f18aef09a4c42bc2143f6d8bdb8179b31318335d0e8e",
  "size": 223,
  "version": 2,
  "locktime": 0,
  "vin": [
    {
      "txid": "2516b9aa394621730eb3acd3c484cbb6904ce4889f0e7a381e01043014fe3542",
      "vout": 0,
      "scriptSig": {
        "asm": "3044022056fe585e8c79441c491c15ca0be9181cb79eea0bbf6f0971fef50d652130e1790220757f32c237289ea846027b9796e33fd97fbb6df6484969800e977dc531b9b483[ALL|FORKID] 03c346eee77a77a8d3e073dacc0532ca7a5b9747bc06d88bf091cac9f4bc8bb792",
        "hex": "473044022056fe585e8c79441c491c15ca0be9181cb79eea0bbf6f0971fef50d652130e1790220757f32c237289ea846027b9796e33fd97fbb6df6484969800e977dc531b9b483412103c346eee77a77a8d3e073dacc0532ca7a5b9747bc06d88bf091cac9f4bc8bb792"
      },
      "sequence": 4294967295
    }
  ],
  "vout": [
    {
      "value": 249.57885442,
      "n": 0,
      "scriptPubKey": {
        "asm": "OP_DUP OP_HASH160 36d2f27bbd826a86db1e93618ce3de89ef331693 OP_EQUALVERIFY OP_CHECKSIG",
        "hex": "76a91436d2f27bbd826a86db1e93618ce3de89ef33169388ac",
        "reqSigs": 1,
        "type": "pubkeyhash",
        "addresses": [
          "bchtest:qqmd9unmhkpx4pkmr6fkrr8rm6y77vckjvqe8aey35"
        ]
      }
    },
    {
      "value": 0.1,
      "n": 1,
      "scriptPubKey": {
        "asm": "OP_HASH160 8024f2cc52373aef482b343bc69b83e6bc031a9c OP_EQUAL",
        "hex": "a9148024f2cc52373aef482b343bc69b83e6bc031a9c87",
        "reqSigs": 1,
        "type": "scripthash",
        "addresses": [
          "bchtest:pzqzfukv2gmn4m6g9v6rh35ms0ntcqc6nsakgt38f8"
        ]
      }
    }
  ],
  "blockhash": "000000006e5de0f297f981bed44fb7cb9dba8bfe6b9133aa3ca551bd61ccf93f",
  "confirmations": 1283,
  "time": 1551712050,
  "blocktime": 1551712050
}
```

**I am trying to spend the n=1 out of this tx.** 

Due to the nature of my project, I am parsing the JSON of the above tx and creating the coins to spend manually using the code below:

_// values from above tx
outAmt = 0.1
scriptPubKeyASM = OP_HASH160 8024f2cc52373aef482b343bc69b83e6bc031a9c OP_EQUAL
txId = "93f7c14688324b0ec296f18aef09a4c42bc2143f6d8bdb8179b31318335d0e8e"
outN = 1_ 

```c#
Mnemonic mnemonic = new Mnemonic(this.Wallet.Mnemonic, Wordlist.English);
ExtKey masterPriv = mnemonic.DeriveExtKey(this.Wallet.Password);
ExtPubKey masterPub = masterPriv.Neuter();
var receivalPrivKey = masterPriv.Derive(this.Wallet.ReceiveChain).Derive(i);

Money txAmt = Money.Parse(outAmt);
Script script = new Script(scriptPubKeyASM);
TxOut txOut = new TxOut(txAmt, script);
OutPoint outPoint = new OutPoint(uint256.Parse(txId), outN);
Coin unspent = new Coin(outPoint, txOut).ToScriptCoin(receivalPrivKey.PrivateKey.ScriptPubKey);
unspentCoins.Add(unspent);
```

**Next I get the change scriptPubKey and create and sign the Transaction**

```c#
// get change addr
ExtPubKey changePub = masterPub.Derive(this.Wallet.ChangeChain).Derive(this.Wallet.NextChangeIndex);
Script changeScriptPubKey = changePub.ScriptPubKey;


// create TX
TransactionBuilder builder = this.Network.CreateTransactionBuilder();
Transaction signedTx = builder
                .AddCoins(coinsToSpend) // coinsToSpend here is the only unspent coin from above
                .AddKeys(signingKeys.ToArray()) // signingKeys in this case contains only the masterPriv value from above
                .Send(toAddress, amountToSend) // amt to send i use 0.05
                .SubtractFees()
                .SetChange(changeScriptPubKey)
                .SendEstimatedFees(new FeeRate(fee)) // fee is usually around 6500~ satoshi
                .BuildTransaction(true);
```

signedTx.Verify() returns true. After this I do signedTx.ToHex() to get the raw transaction and attempt to broadcast the result to the network, but the broadcast fails with the error 16: mandatory-script-verify-flag-failed (signature must be zero for failed check(multi)sig operation)

1. The exact same code worked perfectly fine for BTC and LTC.
2. I am aware that BCash only works with P2SH addresses and am using such addresses
3. I have tired generating an unsigned transaction the same way and signing it using the bitcoin-cli from Bitcoin ABC and broadcasting the result worked fine, which leads me to believe there is an issue with the signing only
4. I am on the latest version of NBitcoin and NBitcoin.Altcoins to date (4.1.1.86 and 1.0.1.38 accordingly)

Please help, I do understand from researching online for a tremendous amount of time, that there are people who have managed to get BCash signing to work, therefore I am pretty sure that there is a logical error in the way I build the TX or addresses, but I am unable to find it. All the help is greatly appreciated. Thank you.

**Comment by NicolasDorier** [comment]
Taking a look right now. Your code seems ok to me.

**Comment by NicolasDorier** [comment]
I tried to reproduce your problem

https://github.com/MetacoSA/NBitcoin/commit/92387a7cefb08e5e50194552412df157c3d91f4a

Sadly, my code pass. Please try to reproduce it, you can have the altcoins tests running with BCash by changing `NodeBuilderEx.cs`.

Another thing you can do is to give me the following

* Alice private key
* The unsigned transaction
* The parent transactions

My wild guess is that the amounts in `coinsToSpend` are incorrect. This can explain why it does not work on BCash (and should not work on segwit either), but works for non segwit stuff.

**Comment by cucaraca** [comment]
Thank you for looking at this Nicolas.

I will investigate your code.

In the meantime, I would like to provide you with some additional info.

- Alice private key
    - By this I assume you mean the private key I use to sign the TX. Here it is:
`cVUn69PYNK7BcbM9fTbwCvcU7rRvS2hbKktoBTiTMkGDJHSWJgXF`
- The unsigned transaction
    - Unsigned RAW tx
`010000000176affc19d751739c22787ec6cbc30d99f9eaf7b7923dea18553d1a9c4c3113740100000000ffffffff02404b4c00000000001976a91468b6b1b42963b427f0a269df0304999bab0511db88ac9a3c4c000000000017a914a5a716c8b6026cd25658b44506624f417ae59edd8700000000`
    - Unsigned tx JSON
```json
{
  "hash": "4c68b14c6ff25253afba84386543a793682f94882db402afae6893a22976b087",
  "ver": 1,
  "vin_sz": 1,
  "vout_sz": 2,
  "lock_time": 0,
  "size": 117,
  "in": [
    {
      "prev_out": {
        "hash": "7413314c9c1a3d5518ea3d92b7f7eaf9990dc3cbc67e78229c7351d719fcaf76",
        "n": 1
      },
      "scriptSig": ""
    }
  ],
  "out": [
    {
      "value": "0.05000000",
      "scriptPubKey": "OP_DUP OP_HASH160 68b6b1b42963b427f0a269df0304999bab0511db OP_EQUALVERIFY OP_CHECKSIG"
    },
    {
      "value": "0.04996250",
      "scriptPubKey": "OP_HASH160 a5a716c8b6026cd25658b44506624f417ae59edd OP_EQUAL"
    }
  ]
}
```

- Parent transaction
     - The vout I am trying to spend is n=1
```json
{
  "hex": "020000000160c32f8c7378d46e8437f49465b67ae64b07db984fbfa3a8c01f4bdfd162832b000000006b483045022100942b8d93fa697fa17a747744e5e6a7cad889699ee6fca187bbed2346842ab91a022079f857383b2159c1c08237dce5c6775b14a1482036c34c15895035e31c66601e412103c346eee77a77a8d3e073dacc0532ca7a5b9747bc06d88bf091cac9f4bc8bb792ffffffff0248e87e55070000001976a91436d2f27bbd826a86db1e93618ce3de89ef33169388ac809698000000000017a914a18d96fe73be11acc399d7e59102e02289f1ff328700000000",
  "txid": "7413314c9c1a3d5518ea3d92b7f7eaf9990dc3cbc67e78229c7351d719fcaf76",
  "hash": "7413314c9c1a3d5518ea3d92b7f7eaf9990dc3cbc67e78229c7351d719fcaf76",
  "size": 224,
  "version": 2,
  "locktime": 0,
  "vin": [
    {
      "txid": "2b8362d1df4b1fc0a8a3bf4f98db074be67ab66594f437846ed478738c2fc360",
      "vout": 0,
      "scriptSig": {
        "asm": "3045022100942b8d93fa697fa17a747744e5e6a7cad889699ee6fca187bbed2346842ab91a022079f857383b2159c1c08237dce5c6775b14a1482036c34c15895035e31c66601e[ALL|FORKID] 03c346eee77a77a8d3e073dacc0532ca7a5b9747bc06d88bf091cac9f4bc8bb792",
        "hex": "483045022100942b8d93fa697fa17a747744e5e6a7cad889699ee6fca187bbed2346842ab91a022079f857383b2159c1c08237dce5c6775b14a1482036c34c15895035e31c66601e412103c346eee77a77a8d3e073dacc0532ca7a5b9747bc06d88bf091cac9f4bc8bb792"
      },
      "sequence": 4294967295
    }
  ],
  "vout": [
    {
      "value": 314.99151432,
      "n": 0,
      "scriptPubKey": {
        "asm": "OP_DUP OP_HASH160 36d2f27bbd826a86db1e93618ce3de89ef331693 OP_EQUALVERIFY OP_CHECKSIG",
        "hex": "76a91436d2f27bbd826a86db1e93618ce3de89ef33169388ac",
        "reqSigs": 1,
        "type": "pubkeyhash",
        "addresses": [
          "bchtest:qqmd9unmhkpx4pkmr6fkrr8rm6y77vckjvqe8aey35"
        ]
      }
    },
    {
      "value": 0.1,
      "n": 1,
      "scriptPubKey": {
        "asm": "OP_HASH160 a18d96fe73be11acc399d7e59102e02289f1ff32 OP_EQUAL",
        "hex": "a914a18d96fe73be11acc399d7e59102e02289f1ff3287",
        "reqSigs": 1,
        "type": "scripthash",
        "addresses": [
          "bchtest:pzscm9h7wwlprtxrn8t7tygzuq3gnu0lxgzxvpj0nt"
        ]
      }
    }
  ],
  "blockhash": "00000000273f1cb02bd590922729b75556e62d3e4b592a812dd19321505a4934",
  "confirmations": 2,
  "time": 1551875543,
  "blocktime": 1551875543
}
```

Yes, in this example `unspentCoins == coinsToSpend` The `outAmt` in `coinsToSpend` is the `value` field amount of the `vout` I am trying to spend of the parent transaction (in this case 0.1)

I also uploaded the full method source [here](https://codeshare.io/5NQrXV) for you to take a look:

**Comment by NicolasDorier** [comment]
Please give me the parent transaction as hex

**Comment by NicolasDorier** [comment]
I also assume your Network instance is set to `BCash.Instance.Regtest` right?

**Comment by cucaraca** [comment]
Hex of the parent tx: `020000000160c32f8c7378d46e8437f49465b67ae64b07db984fbfa3a8c01f4bdfd162832b000000006b483045022100942b8d93fa697fa17a747744e5e6a7cad889699ee6fca187bbed2346842ab91a022079f857383b2159c1c08237dce5c6775b14a1482036c34c15895035e31c66601e412103c346eee77a77a8d3e073dacc0532ca7a5b9747bc06d88bf091cac9f4bc8bb792ffffffff0248e87e55070000001976a91436d2f27bbd826a86db1e93618ce3de89ef33169388ac809698000000000017a914a18d96fe73be11acc399d7e59102e02289f1ff328700000000`

Network is `BCash.Instance.Testnet`

**Comment by rocketfuelblockchain** [comment]
Did we reach a resolution?  I am having a very similar problem.  TransactionBuilder.Verify loves my transaction and passes OK.  But BTrash network won't swallow it when pushed via btc.com.  I wonder if TransactionBuilder is imperfect?  Shall I post up some code?

**Comment by cucaraca** [comment]
No, we have not reached a solution, I am still waiting for Nicolas to reply now after providing him the last few bits of info. 

As soon as I figure out myself if Nicolas does not find anything I'll let you know. But I believe you should post anything you think might help. If I am not alone having this issue it might actually end up being a bug in the library code for a specific scenario and I feel we should provide all the info we can to help pinpoint the issue.

Also, if you find a solution yourself please post a response here. I would really appreciate any help I can get :)

**Comment by rocketfuelblockchain** [comment]
OK cucaraca - I'll kick it around a bit more.  If I get anything I'll post it here.  Meanwhile, I'll also be waiting wait for the master to get back to us.  If I can get mine to work, I'll post up a sample tomorrow.

**Comment by rocketfuelblockchain** [comment]
Sorry for the VB boys and girls!!!  But here is an example of Transaction Builder saying yes when network won't process it.  Not sure what I did wrong.

>         Dim ntwk As Network = BCash.Instance.Mainnet
>         Dim bcaSource As BitcoinAddress = BitcoinAddress.Create("bitcoincash:qp7xpt2gss0x9d9m6fuzs78lv05sdzw22u0g4rsx8w", ntwk)
>         Dim bcaPayTo As BitcoinAddress = BitcoinAddress.Create("bitcoincash:qryparnm0l72qsas3z5ey72mzky8e9s4jgts8a2tsg", ntwk)
>         Dim bcsSource As BitcoinSecret = New BitcoinSecret("L3QN26rV4Pt8cAmRb72g43JM6RAKKngnMcAKg5GC9T7UyfA4gTu3")
> 
>         Dim txBuild = ntwk.CreateTransactionBuilder
> 
>         Dim mCoin As New Coin
>         mCoin.Amount = Money.Satoshis(3136271)
>         mCoin.Outpoint = New OutPoint(uint256.Parse("49fc00819134b956c0f69319300b3ee5bc764887b16361a736153c0a139abf27"), 1)
>         mCoin.ScriptPubKey = bcsSource.ScriptPubKey
>         txBuild.AddCoins(mCoin) 'add the funding coins
> 
>         txBuild.AddKeys(bcsSource) 'add the signing keys 
> 
>         txBuild.Send(bcaPayTo, Money.Satoshis(3000000)) 'payTo address; amount
>         txBuild.Send(TxNullDataTemplate.Instance.GenerateScriptPubKey(Encoding.UTF8.GetBytes("Thanks NBitcoin!")), Money.Zero) 'OPReturn
>         txBuild.SendFees(Money.Satoshis(10000)) 'network fee
>         txBuild.SetChange(bcaSource) 'Send the change back to the wallet address
> 
>         Dim tx As Transaction = txBuild.BuildTransaction(True)
>         If txBuild.Verify(tx) Then
>             MsgBox("Its all good", MsgBoxStyle.OkOnly)
>         Else
>             MsgBox("Fucked Up", MsgBoxStyle.OkOnly)
>         End If

Transaction in Hex: 
`010000000127bf9a130a3c1536a76163b1874876bce53e0b301993f6c056b934918100fc49010000006a47304402205f89f25f02b2a7723308623a0af7466f64d151646d334b2deab9ac50c8d1fd67022023a073d77d09a2ac72a27cfb14504fa49879f05f6624c8e42d41378e75bc6c50412103c63a085c54a1dfc1c2f49b375ee64ed0585cfee71fd8627e1d2c6eeaa672a098ffffffff03c0c62d00000000001976a914c81e8e7b7ffca043b088a992795b15887c96159288ac0000000000000000126a105468616e6b73204e426974636f696e213fed0100000000001976a9147c60ad48841e62b4bbd2782878ff63e90689ca5788ac00000000`

```
{{
  "hash": "60a40d9ecb0a30b0ee445b711bf3947ef13c5e697986d896b064bf406554216b",
  "ver": 1,
  "vin_sz": 1,
  "vout_sz": 3,
  "lock_time": 0,
  "size": 252,
  "in": [
    {
      "prev_out": {
        "hash": "49fc00819134b956c0f69319300b3ee5bc764887b16361a736153c0a139abf27",
        "n": 1
      },
      "scriptSig": "304402205f89f25f02b2a7723308623a0af7466f64d151646d334b2deab9ac50c8d1fd67022023a073d77d09a2ac72a27cfb14504fa49879f05f6624c8e42d41378e75bc6c5041 03c63a085c54a1dfc1c2f49b375ee64ed0585cfee71fd8627e1d2c6eeaa672a098"
    }
  ],
  "out": [
    {
      "value": "0.03000000",
      "scriptPubKey": "OP_DUP OP_HASH160 c81e8e7b7ffca043b088a992795b15887c961592 OP_EQUALVERIFY OP_CHECKSIG"
    },
    {
      "value": "0.00000000",
      "scriptPubKey": "OP_RETURN 5468616e6b73204e426974636f696e21"
    },
    {
      "value": "0.00126271",
      "scriptPubKey": "OP_DUP OP_HASH160 7c60ad48841e62b4bbd2782878ff63e90689ca57 OP_EQUALVERIFY OP_CHECKSIG"
    }
  ]
}}
```

It seems like that should work to me.  But https://btc.com/tools/tx/publish  won't push it.

**Comment by rocketfuelblockchain** [comment]
Oooops.  btc.com is pushing to BTC chain, not BCH chain.  Not sure why they don't offer a BCH push API/Tool.  Anyway, the transaction was fine and did runon BCH main chain as planned.  So, my problem went away for very dumb reason.  :)

**Comment by NicolasDorier** [comment]
@cucaraca sorry, I did not see your reply. Taking a look.

**Comment by NicolasDorier** [comment]
@cucaraca 

```
var network = Altcoins.BCash.Instance.Testnet;
var aliceKey = network.CreateBitcoinSecret("cVUn69PYNK7BcbM9fTbwCvcU7rRvS2hbKktoBTiTMkGDJHSWJgXF");

Transaction unsignedTransaction = Transaction.Parse("010000000176affc19d751739c22787ec6cbc30d99f9eaf7b7923dea18553d1a9c4c3113740100000000ffffffff02404b4c00000000001976a91468b6b1b42963b427f0a269df0304999bab0511db88ac9a3c4c000000000017a914a5a716c8b6026cd25658b44506624f417ae59edd8700000000", network);
Transaction parentTransaction = Transaction.Parse("020000000160c32f8c7378d46e8437f49465b67ae64b07db984fbfa3a8c01f4bdfd162832b000000006b483045022100942b8d93fa697fa17a747744e5e6a7cad889699ee6fca187bbed2346842ab91a022079f857383b2159c1c08237dce5c6775b14a1482036c34c15895035e31c66601e412103c346eee77a77a8d3e073dacc0532ca7a5b9747bc06d88bf091cac9f4bc8bb792ffffffff0248e87e55070000001976a91436d2f27bbd826a86db1e93618ce3de89ef33169388ac809698000000000017a914a18d96fe73be11acc399d7e59102e02289f1ff328700000000", network);
var txBuilder = network.CreateTransactionBuilder();
txBuilder.AddCoins(parentTransaction.Outputs.AsCoins());
txBuilder.AddKeys(aliceKey);
var signed = txBuilder.SignTransactionInPlace(unsignedTransaction);
Assert.True(txBuilder.Verify(signed));
```

The transaction I get is:

```json
{
  "hash": "756c9459c53db33673e5272f93741b3c4e72feff4377cda8c96016ac504664d2",
  "ver": 1,
  "vin_sz": 1,
  "vout_sz": 2,
  "lock_time": 0,
  "size": 249,
  "in": [
    {
      "prev_out": {
        "hash": "7413314c9c1a3d5518ea3d92b7f7eaf9990dc3cbc67e78229c7351d719fcaf76",
        "n": 1
      },
      "scriptSig": "304402200ad6eba45ef11d19935aa9b15400aa75b59e8f2badf66c845588f524fe23cb97022050c3dec0c39bf978af2729b1701a51d546648bb7761381aa300c2a62a4df1a7041 02d3d236142bae196983f41ac20651214e5dff28b0644fe28ee11706225dd59a20 76a914e8841dc0a61a560dee1440a0eab4edc653c4874d88ac"
    }
  ],
  "out": [
    {
      "value": "0.05000000",
      "scriptPubKey": "OP_DUP OP_HASH160 68b6b1b42963b427f0a269df0304999bab0511db OP_EQUALVERIFY OP_CHECKSIG"
    },
    {
      "value": "0.04996250",
      "scriptPubKey": "OP_HASH160 a5a716c8b6026cd25658b44506624f417ae59edd OP_EQUAL"
    }
  ]
}
```

or in hex 
```
010000000176affc19d751739c22787ec6cbc30d99f9eaf7b7923dea18553d1a9c4c311374010000008447304402200ad6eba45ef11d19935aa9b15400aa75b59e8f2badf66c845588f524fe23cb97022050c3dec0c39bf978af2729b1701a51d546648bb7761381aa300c2a62a4df1a70412102d3d236142bae196983f41ac20651214e5dff28b0644fe28ee11706225dd59a201976a914e8841dc0a61a560dee1440a0eab4edc653c4874d88acffffffff02404b4c00000000001976a91468b6b1b42963b427f0a269df0304999bab0511db88ac9a3c4c000000000017a914a5a716c8b6026cd25658b44506624f417ae59edd8700000000
```

**Comment by NicolasDorier** [comment]
@cucaraca 

![image](https://user-images.githubusercontent.com/3020646/54280211-4687c200-45da-11e9-9a88-68db99b4dd58.png)

**Comment by NicolasDorier** [comment]
@cucaraca 

Can you check if `signed.GetType()` is a `ForkIdTransaction`? (as well as all transactions you are manipulating)

**Comment by cucaraca** [comment]
Sorry for the late response, and thank you for coming back to me on this Nicolas. Yes `signed.GetType()` is actually `ForkIdTransaction` and I am not manipulating any other `Transaction` objects. I am manually parsing the tx json returned by the RPC to create the `Coin` objects. Is there any chance there is a different type for `Coin` as well that i should be using when working with BCH?

I will try to replicate your code and parse the transaction the same way you are doing in the comment above. Maybe the unspent outputs and unspent coins are compiled somewhat differently using your suggested approach. I will let you know as soon as I have some info.

Thanks again.

**Comment by NicolasDorier** [comment]
@cucaraca the only problem I can see which would match your issue is if the coin you are spending have the wrong `Value`.

**Comment by cucaraca** [comment]
@NicolasDorier I am completely lost at this point...

First of all to reply to your latest comment - the coin `Value` cannot really be incorrect since I am getting that directly from the RPC. Also if I use the same transaction and feed it to the RPC to the signrawtransaction call it signs and works fine. Moreover - you have successfully used the hex of the unsigned transaction that I generated and you managed to sign and broadcast it. The `Value` used in that tx you signed was the same I use and was received in the same way.

**Now to the real head-scratcher:**

I have replicated your code 100% [from this comment](https://github.com/MetacoSA/NBitcoin/issues/664#issuecomment-472407821) and I do get the same `signed` hex if I use the same `unsignedTransaction` hex, same `parentTransaction` hex and same private key for `aliceKey` you used in your comment above. This tells me that both our codes run the same way, that I am not on a different buggy branch of `Altcoins,` or something like that.

Then I used this same code filled with values from a new transaction I received 

```c#
Transaction unsignedTransaction = Transaction.Parse("0100000001b883962a72ad96f81e6775c427db9ea1c311d4010353e2a584417549fa3ebf980100000000ffffffff02d03c8900000000001976a9148cd1093ba274738c5c73958995de4362a074e66488ac40420f000000000017a91413951668c9b017e1d503690c9117acf323548a7b8700000000", network);
Transaction parentTransaction = Transaction.Parse("0200000001563e59bda1e7f9627fd78f64b2c8a53425f6ef9540a97c818a50dd3abff2afb2000000006b483045022100cba1ea4682b8ffcae0ee5192f38da13e23656778664134887b49856d70ad557502205d73a5249c883e51bfd996a6a440bcaac5687624be00867f091e62b0a854e03e412103c346eee77a77a8d3e073dacc0532ca7a5b9747bc06d88bf091cac9f4bc8bb792ffffffff0212915301070000001976a91436d2f27bbd826a86db1e93618ce3de89ef33169388ac809698000000000017a914b48513ff6314d81ca64bc43e029f3fd9137414e38700000000", network);
var txBuilder = network.CreateTransactionBuilder();
txBuilder.AddCoins(parentTransaction.Outputs.AsCoins());
txBuilder.AddKeys("cQBTYw1CggvM8h8u1VCgiqaGdzhgsmDzhint65wJ3rjzBmWNVMB4");
var signed = txBuilder.SignTransaction(unsignedTransaction);
```

The `signed` hex value I get
`0100000001b883962a72ad96f81e6775c427db9ea1c311d4010353e2a584417549fa3ebf98010000008447304402205709445db39a273e45686e3a762e2566640b40ebfddb79da618d88deaa9b5865022011905d4e82bf517499bd29803eb899f1ed0315f29222119a847518fb883801f04121020bc2b4cdd9334ccd5676137ea19149c3e6d9259c2ba28ac1ee99a5cb91e15b781976a9147ffb4aa7f5018b4764f09eb66b42ace78a4c507088acffffffff02d03c8900000000001976a9148cd1093ba274738c5c73958995de4362a074e66488ac40420f000000000017a91413951668c9b017e1d503690c9117acf323548a7b8700000000`

However, trying to submit it to the network does not work...

![image](https://user-images.githubusercontent.com/48249248/54528216-97057380-497c-11e9-9dd5-e26ffd8faf78.png)

Based on my understanding you should be getting the same `signed` hex value. So if you manage to broadcast it what is different here? Are you broadcasting it somewhat differently? I am using this url to broadcast my transactions while testing https://tbch.blockdozer.com/tx/send

Another thing - when I try to sign the transaction using the RPC command - the transaction signs successfully but the output hex is different. I have not tried broadcastng this transaction, but based on my previous experience this signed tx would work. I dont want to spend this TX so that you could try the values yourself. 

I make this call to the bitcoin CLI for BitcoinABC to sign the tx:
`"C:\Program Files\Bitcoin-abc\daemon\bitcoin-cli" -rpcuser=user -rpcpassword=pass -rpcport=port -testnet signrawtransaction 0100000001b883962a72ad96f81e6775c427db9ea1c311d4010353e2a584417549fa3ebf980100000000ffffffff02d03c8900000000001976a9148cd1093ba274738c5c73958995de4362a074e66488ac40420f000000000017a91413951668c9b017e1d503690c9117acf323548a7b8700000000 "[{\"txid\":\"98bf3efa49754184a5e2530301d411c3a19edb27c475671ef896ad722a9683b8\",\"vout\":1,\"scriptPubKey\":\"a914b48513ff6314d81ca64bc43e029f3fd9137414e387\",\"redeemScript\":\"76a9147ffb4aa7f5018b4764f09eb66b42ace78a4c507088ac\",\"amount\":\"0.01000000\"}]" "[\"cQBTYw1CggvM8h8u1VCgiqaGdzhgsmDzhint65wJ3rjzBmWNVMB4\"]" "ALL|FORKID"`

this gives me the following hex
`0100000001b883962a72ad96f81e6775c427db9ea1c311d4010353e2a584417549fa3ebf9801000000844730440220612d4e9480fa98ee630a81851bb9dd00fd3f5857b4116c79f8385d94898be98b02205dbfd9967ab2213e96a6fac3de645fc86c7e4c6e7eeba5e766b27c85a1b4387b4121020bc2b4cdd9334ccd5676137ea19149c3e6d9259c2ba28ac1ee99a5cb91e15b781976a9147ffb4aa7f5018b4764f09eb66b42ace78a4c507088acffffffff02d03c8900000000001976a9148cd1093ba274738c5c73958995de4362a074e66488ac40420f000000000017a91413951668c9b017e1d503690c9117acf323548a7b8700000000`

Nicolas, I would be willing to communicate with you directly via Skype or email for more efficient communication until we pinpoint the issue. If you feel this is a good idea, feel free to email me at anoncucaracha@gmail.com

Thank you for your time and input.

**Comment by NicolasDorier** [comment]
@cucaraca can you come on https://chat.btcpayserver.org ? we have a NBitcoin chat there. Taking a look at your transaction now.

**Comment by NicolasDorier** [comment]
And I don't get the same value as you by copy pasting your code

**Comment by NicolasDorier** [comment]
This is what I get as signed transaction

```
0100000001b883962a72ad96f81e6775c427db9ea1c311d4010353e2a584417549fa3ebf98010000008447304402200997b3e3c84be59a28cf9028a3ba338d42465f6d88fad5f2bbe8d93071568ea202207c54b4de9b70ffb3d8ed90805ee0c20cb0654a95252bab5e8387ab0ac61487124121020bc2b4cdd9334ccd5676137ea19149c3e6d9259c2ba28ac1ee99a5cb91e15b781976a9147ffb4aa7f5018b4764f09eb66b42ace78a4c507088acffffffff02d03c8900000000001976a9148cd1093ba274738c5c73958995de4362a074e66488ac40420f000000000017a91413951668c9b017e1d503690c9117acf323548a7b8700000000
```

**Comment by NicolasDorier** [comment]
The transaction I get you get from bitcoin core might be different because of sig malleability. Trying to broadcast myt x...

**Comment by NicolasDorier** [comment]
![image](https://user-images.githubusercontent.com/3020646/54536673-7b2dbc00-49d4-11e9-8da8-9e3ea6a61e3a.png)

**Comment by NicolasDorier** [comment]
I literally generated this transaction by copy and pasting your code.

What environment are you running on?

**Comment by NicolasDorier** [comment]
Your library (NBitcoin and NBitcoin.Altcoins) are up to dates?

**Comment by cucaraca** [comment]
Running on .NET Framework on Windows. 
NBitcoin version 4.1.1.86 
NBitcoin.Altcoins version 1.0.1.38

Coming to your chat now

**Comment by cucaraca** [comment]
SOLVED!!!

I think this is very unlikely that anyone will ever come across a similar scenario but not to leave this hanging - here goes.

I'll try to explain the best i can.
The project that was failing the signing is a large project that is a long time in development. It currently is structured as follows: 
It has a wpf `UI` project that holds the interface logic for the app and it has the core logic (lets call it `backend`) project that contains all the crypto logic within.
The `UI` project is referencing the `backend` project and calling its functions.

The problem became apparent when running the same code on Nicolas' machine and mine started yileding different results. I made sure the `backend` project was referencing the correct NBitcoin version etc. however I never looked at the `UI` project. And why would I, it should not even matter.

Turns out it did matter. At some point along the dev cycle, NBitcoin reference was added to the **UI** project. The logic was since then removed or moved to the `backend` project but the reference was not removed. So the `UI` project had a reference to NBitcoin 4.1.1.67 (old version) while `backend` was referencing the propper new version of NBitcoin. However for some reason when the `backend` project functions were being called from the `UI` project the NBitcoin reference that was doing the calculations was the old one from the `UI` project. After removing the outdated NBitcoin reference from the `UI` project the `backend` project started working and signing fine.

I am still having hard time understanding why this would be. This might be a by design functionality of .net might be a bug. Not sure, but it is what it is.

Just want to thank again to Nicolas for being so hands on and responsive with this issue.
